if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("tximport",update = FALSE,ask = FALSE)
BiocManager::install("DESeq2",update = FALSE,ask = FALSE)

install.packages(c('pheatmap','mclust','reshape2','ggplot2','readr'))


library(DESeq2)
library(tximport)
library(ggplot2)
library(pheatmap)
library(reshape2)

#Delete everithing in your R environment
rm(list=ls())

#Set your working directory. This should be the directory containing all the Salmon results, remember you should have 16 folders with salmon results
wd<-"/data/diriano/cen5789_salmon/"
setwd(wd)

#Loading the experimental design
#we use the R read.delim to load a text file, that has columns separated by TABs, this is stores inside R as an object called target, a data.frame
targets<-read.delim("targets.txt",header=T)
rownames(targets)<-targets$SampleName
targets$Genotype<-as.factor(targets$Genotype)
targets$EnvironmentalStress<-as.factor(targets$EnvironmentalStress)

#Preparing to load the salmon quantification
#We need to prepare and object that has the sample names, and the location on disk of the quant.sf files produced by salmon
#We have the samples names in the target object, in the SampleName column,
# and we know that in our working directoy we have a folder for each sample,
# with the same name as the SampleName and the quat.sf file is inside that directory

myFiles<-paste(wd, targets$SampleName,"/quant.sf",sep="")

#Check whether these paths look ok
myFiles
#Add the sample name as the name of each path
names(myFiles)<-targets$SampleName
myFiles

#Check whether these paths actually exists on disk
all(file.exists(myFiles))

#Now, let load the mapping of transcript to gene identifiers
tx2gene<-read.delim("tx2gene.txt",header=F)
head(tx2gene)

#Let's see one example of a gene that has 3 transcripts
tx2gene[which(tx2gene$V2=='AT1G51370'),]

#Now, let's import the salmon quantification data into R, for this we will use the package tximport
txi.salmon<-tximport(files = myFiles, type = 'salmon', tx2gene = tx2gene, txIn = TRUE, txOut = FALSE)

#Just in order to compare between the quantification at the level of gene and transcript we will load the data at the level of transcripts
txi.salmon.tx<-tximport(files = myFiles, type = 'salmon', tx2gene = tx2gene, txIn = TRUE, txOut = TRUE)
head(txi.salmon$counts['AT1G51370',])
head(txi.salmon.tx$counts[c('AT1G51370.1','AT1G51370.2','AT1G51370.3'),])
#Let's just remove the quantification at the level of transcrip, as we will not use it
rm(txi.salmon.tx)

seqEffort<-as.data.frame(colSums(txi.salmon$counts))
seqEffort$Samples<-rownames(seqEffort)
colnames(seqEffort)<-c('NumberFragments', 'Samples')
#Let's compare the sequencing effort for all samples
p<- ggplot(seqEffort, aes(x=Samples,y=NumberFragments))+ 
  geom_col()

# Let's try to improve the figure, remove the background and put the sample names at an angle
p + theme_bw() +
  theme(axis.text.x = element_text(angle=60, size = 12, hjust = 1))
# Let's change the axis names
p + theme_bw() +
  theme(axis.text.x = element_text(angle=60, size = 12, hjust = 1)) +
  xlab("Amostras") +
  ylab("Número de fragmentos sequenciados")

#We will carry out the identification of DEG using the package DESeq2, so let's import our raw counts into DESeq2
targets$Number<-NULL
targets$SampleName<-NULL
dds <- DESeqDataSetFromTximport(txi.salmon, colData = targets, design = ~Genotype + EnvironmentalStress)

#Define the base level condition for the experimental factors under study
dds$Genotype <-relevel(dds$Genotype, ref='WT')
dds$EnvironmentalStress <-relevel(dds$EnvironmentalStress, ref='None')

# Let's have a loot at the object generated by DESeq2
head(dds)
assay(dds)
colData(dds)
rowData(dds)
dimnames(dds)
metadata(dds)

counts_melt<-melt(assay(dds))


colnames(counts_melt)<-c("Gene","Sample","Count")
ggplot(counts_melt,aes(x=Sample,y=Count)) + 
  geom_boxplot()+
  scale_y_log10()+
  theme_bw() +
  theme(axis.text.x = element_text(angle=60, size = 12, hjust = 1)) +
  xlab("Amostras") +
  ylab("Número de fragmentos sequenciados")

keep <- rowSums(assay(dds) >= 1) >= 1

table(keep)
dim(dds)
dds <- dds[keep,]
dim(dds)

dds <- DESeq(dds)

vsd <- vst(dds, blind=TRUE)
targets
plotPCA(vsd, intgroup=c("Genotype"),ntop=500)  + theme_bw()
plotPCA(vsd, intgroup=c("EnvironmentalStress"),ntop=500) + theme_bw()
plotPCA(vsd, intgroup=c("EnvironmentalStress", "Genotype"),ntop=500)  + theme_bw()

sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$EnvironmentalStress, vsd$Genotype, sep="-")
colnames(sampleDistMatrix) <- paste(vsd$EnvironmentalStress, vsd$Genotype, sep="-")
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists)


res_SALT_vs_Control <- results(dds, lfcThreshold=2, altHypothesis="greaterAbs", contrast = c('EnvironmentalStress','NaCl','None'), alpha=0.05)
res_ABA_vs_Control <- results(dds, lfcThreshold=2, altHypothesis="greaterAbs", contrast = c('EnvironmentalStress','ABA','None'), alpha=0.05)
res_Drought_vs_Control <- results(dds, lfcThreshold=2, altHypothesis="greaterAbs", contrast = c('EnvironmentalStress','Drought','None'), alpha=0.05)


drawLines <- function() abline(h=c(-2,2),col="dodgerblue",lwd=2)
plotMA(res_SALT_vs_Control, main = "SALT_vs_Control"); drawLines()
plotMA(res_ABA_vs_Control, main = "ABA_vs_Control"); drawLines()
plotMA(res_Drought_vs_Control, main = "Drought_vs_Control"); drawLines()

df_res_SALT_vs_Control<-as.data.frame(res_SALT_vs_Control)
df_res_SALT_vs_Control$diffExpressed <- "NO"
df_res_SALT_vs_Control$diffExpressed[df_res_SALT_vs_Control$padj < 0.05 & df_res_SALT_vs_Control$log2FoldChange >0] <- "UP"
df_res_SALT_vs_Control$diffExpressed[df_res_SALT_vs_Control$padj < 0.05 & df_res_SALT_vs_Control$log2FoldChange <0] <- "DOWN"
table(df_res_SALT_vs_Control$diffExpressed)

df_res_ABA_vs_Control<-as.data.frame(res_ABA_vs_Control)
df_res_ABA_vs_Control$diffExpressed <- "NO"
df_res_ABA_vs_Control$diffExpressed[df_res_ABA_vs_Control$padj < 0.05 & df_res_ABA_vs_Control$log2FoldChange >0] <- "UP"
df_res_ABA_vs_Control$diffExpressed[df_res_ABA_vs_Control$padj < 0.05 & df_res_ABA_vs_Control$log2FoldChange <0] <- "DOWN"
table(df_res_ABA_vs_Control$diffExpressed)

df_res_Drought_vs_Control<-as.data.frame(res_Drought_vs_Control)
df_res_Drought_vs_Control$diffExpressed <- "NO"
df_res_Drought_vs_Control$diffExpressed[df_res_Drought_vs_Control$padj < 0.05 & df_res_Drought_vs_Control$log2FoldChange >0] <- "UP"
df_res_Drought_vs_Control$diffExpressed[df_res_Drought_vs_Control$padj < 0.05 & df_res_Drought_vs_Control$log2FoldChange <0] <- "DOWN"
table(df_res_Drought_vs_Control$diffExpressed)

ggplot(df_res_SALT_vs_Control, aes(x=log2FoldChange, y=-log10(padj), col=diffExpressed)) + 
  theme_bw() +
  geom_point() +
  geom_vline(xintercept = c(-2, 2), col = "gray", linetype = 'dashed')+
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed')+
  scale_color_manual(values = c("#00AFBB", "grey", "#FFDB6D"), 
                     labels = c("Downregulated", "Not significant", "Upregulated"))+
  ggtitle('SALT_vs_Control DEGs')+
  ylim(0,3)


pheatmap(assay(vsd)[row.names(assay(vsd)) %in%
                      row.names(df_res_SALT_vs_Control[which(df_res_SALT_vs_Control$diffExpressed %in% 
                                                               c('UP','DOWN')),]),
                    row.names(targets[which(targets$EnvironmentalStress %in% c("NaCl","None")),])],
         scale='row', 
         annotation_col = targets,
         main = "SALT_vs_Control DEGs")

pheatmap(assay(vsd)[row.names(assay(vsd)) %in%
                      row.names(df_res_ABA_vs_Control[which(df_res_ABA_vs_Control$diffExpressed %in% 
                                                               c('UP','DOWN')),]),
                    row.names(targets[which(targets$EnvironmentalStress %in% c("ABA","None")),])],
         scale='row', 
         annotation_col = targets,
         main = "ABA_vs_Control DEGs")

pheatmap(assay(vsd)[row.names(assay(vsd)) %in%
                      row.names(df_res_Drought_vs_Control[which(df_res_Drought_vs_Control$diffExpressed %in% 
                                                               c('UP','DOWN')),]),
                    row.names(targets[which(targets$EnvironmentalStress %in% c("Drought","None")),])],
         scale='row', 
         annotation_col = targets,
         main = "Drought_vs_Control DEGs")

plotCounts(dds, gene=which.min(res_ABA_vs_Control$padj), intgroup="EnvironmentalStress", normalized = TRUE)
plotCounts(dds, gene=which.min(res_Drought_vs_Control$padj), intgroup="EnvironmentalStress", normalized = TRUE)
plotCounts(dds, gene=which.min(res_SALT_vs_Control$padj), intgroup="EnvironmentalStress", normalized = TRUE)

write.table(df_res_SALT_vs_Control[which(df_res_SALT_vs_Control$diffExpressed %in% c('UP','DOWN')),],
            file = "SALT_vs_Control_DEGs.txt", sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE)

write.table(df_res_ABA_vs_Control[which(df_res_ABA_vs_Control$diffExpressed %in% c('UP','DOWN')),],
            file = "ABA_vs_Control_DEGs.txt", sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE)

write.table(df_res_Drought_vs_Control[which(df_res_Drought_vs_Control$diffExpressed %in% c('UP','DOWN')),],
            file = "Drought_vs_Control_DEGs.txt", sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE)
